### PR TITLE
#4 - Add test summary

### PR DESCRIPTION
**Pull request by bhavaniravi**
Adds #3 and #4

Add a test summary at the end of the test suite.

I've attempted this to the best of my knowledge. Any suggestions, comments, improvements are welcomed

### CODE REVIEW

**Comment by nicoddemus** [comment]
Here's the output:

image

File: pytest_rich.py [hunk 1]
```diff
@@ -243,8 +244,48 @@ def pytest_sessionfinish(
...
+                        Text("SUCCESS ", style="green")
+                        + Text(f"{nodeid}", style="black")
+                    )
+
+            self.console.print(Rule(characters="-", style="blue"))
+
+        for nodeid, errors in error_messages.items():
+            self.console.print(
+                Text("FAILED ", style="red")
+                + Text(f"{nodeid} {''.join(errors)}", style="black")
```
Comment (nicoddemus) [inline pytest_rich.py:284]: Ditto here for black style

File: pytest_rich.py [hunk 2]
```diff
@@ -243,8 +244,48 @@ def pytest_sessionfinish(
...
+            )
+
+        self.console.print(Rule("The End", characters="=", style="blue"))
```
Comment (nicoddemus) [inline pytest_rich.py:287]: This is of course a matter of taste, but I don't think we need this rule at the end of the session... if we do want to have it however, we might display the exit status instead (given by `pytest_sessionfinish`).

**Review by nicoddemus — CHANGES_REQUESTED** [review]
Thanks @bhavaniravi overall looks nice, left a few request for changes though.

Also, can you please update the changelog? Thanks!

File: pytest_rich.py [hunk 3]
```diff
@@ -243,8 +244,48 @@ def pytest_sessionfinish(
...
+        total_text = Text(f" Total::{self.total_items_completed}", style="bold blue")
+        success_text = Text(
+            f" Success::{self.total_items_completed - len(self.failed_reports)}",
+            style="bold green",
+        )
+        failed_text = Text(f" Failed::{len(self.failed_reports)}", style="bold red")
+        self.console.print(
+            Rule(
+                Text("Short test summary info")
+                + total_text
```
[the remark this answers is not in the document]
Comment (nicoddemus) [reply pytest_rich.py:263]: A few suggestions, leaving them here to see what @joshuadavidthomas thinks:

1. We might consider changing `::` to `:`, so it reads `Total: 38 Success: 38 Failed: 0`.
2. We should be "smart" like pytest is, for example, if all tests pass, it shows only `64 passed`. If some failed, it shows `60 passed, 4 failed` and so on. So here we could use `Success: 38` if all passed, but show `Total: 38 Success: 30 Failed: 8` if some fail, for example.
3. We are not showing the time taken by the suite, which is often useful. We might leverage rich's abilities here and instead show a box with the summary. For example here's how [ward ](https://github.com/darrenburns/ward) does it: 
image

File: pytest_rich.py [hunk 4]
```diff
@@ -243,8 +244,44 @@ def pytest_sessionfinish(
...
+            style="bold green",
+        )
+        failed_text = Text(f" Failed::{len(self.failed_reports)}", style="bold red")
+        self.console.print(
+            Rule(total_text
+                + success_text
+                + failed_text,
+                characters="=",
+                style="blue",
+            )
```
Comment (joshuadavidthomas) [inline pytest_rich.py:266]: Spacing is a bit off with the spaces in the total and failed `Text` objects:

image

Could you change this to adding the spacing when printing to the console?

```python
total_text = Text(f"Total::{self.total_items_completed}", style="bold blue")
success_text = Text(
	f"Success::{self.total_items_completed - len(self.failed_reports)}",
    	style="bold green",
	)
failed_text = Text(f"Failed::{len(self.failed_reports)}", style="bold red")
self.console.print(
    Rule(
        total_text + " " + success_text + " " + failed_text,
        characters="=",
        style="blue",
    )
)
```

image

**Review by joshuadavidthomas — CHANGES_REQUESTED** [review]
This is great @bhavaniravi! Sorry for the silence the past couple of months regarding your contribution. I've left some comments on my review and @nicoddemus's.

File: pytest_rich.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (joshuadavidthomas) [reply pytest_rich.py]: @nicoddemus Yeah, I think the time would be useful, but may be worth tackling in a separate PR.

As far as the colons, yeah I agree, single is clearer.

**Comment by bhavaniravi** [comment]
Currently looks like this

**All Pass**

image

**Failures**

image

**Dark Mode**

image

**Review by joshuadavidthomas — APPROVED** [review]
LGTM, thanks!

**Pull request merged**